File: weather-demo-app-apple/Weather/App/WeatherAppRoot.swift
//
//  WeatherAppRoot.swift
//  Weather
//
//  Created by Onur on 4/11/26.
//

import SwiftUI

struct WeatherAppRoot: View {
    @State private var path: [Route] = []

    @ObservedObject var welcomeViewModel: WelcomeViewModel
    @ObservedObject var currentWeatherViewModel: CurrentWeatherViewModel
    @ObservedObject var forecastViewModel: ForecastViewModel

    var body: some View {
        NavigationStack(path: $path) {
            Group {
                if welcomeViewModel.hasSeenWelcome {
                    CurrentWeatherScreen(
                        viewModel: currentWeatherViewModel,
                        onOpenForecast: { city in
                            path.append(.forecast(city: city))
                        }
                    )
                } else {
                    WelcomeScreen(
                        viewModel: welcomeViewModel,
                        onContinue: {
                            path = []
                        }
                    )
                }
            }
            .navigationDestination(for: Route.self) { route in
                switch route {
                case .forecast(let city):
                    ForecastScreen(
                        viewModel: forecastViewModel,
                        city: city
                    )

                default:
                    EmptyView()
                }
            }
        }
    }
}

File: weather-demo-app-android/app/src/main/java/com/demo/weather/app/WeatherAppRoot.kt
package com.demo.weather.app

import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.navigation.compose.NavHost
import androidx.navigation.compose.composable
import androidx.navigation.compose.rememberNavController
import androidx.navigation.toRoute
import com.demo.weather.ui.current.CurrentWeatherScreen
import com.demo.weather.ui.current.CurrentWeatherViewModel
import com.demo.weather.ui.forecast.ForecastScreen
import com.demo.weather.ui.forecast.ForecastViewModel
import com.demo.weather.ui.welcome.WelcomeScreen
import com.demo.weather.ui.welcome.WelcomeViewModel

@Composable
fun WeatherAppRoot(
    welcomeViewModel: WelcomeViewModel,
    currentWeatherViewModel: CurrentWeatherViewModel,
    forecastViewModel: ForecastViewModel
) {
    val hasSeenWelcome by welcomeViewModel.hasSeenWelcome.collectAsState()
    val startDestination = if (hasSeenWelcome) Route.Current else Route.Welcome

    val navController = rememberNavController()

    NavHost(
        navController = navController,
        startDestination = startDestination
    ) {
        composable<Route.Welcome> {
            WelcomeScreen(
                viewModel = welcomeViewModel,
                onContinue = {
                    navController.navigate(Route.Current) {
                        popUpTo(Route.Welcome) { inclusive = true }
                    }
                }
            )
        }

        composable<Route.Current> {
            CurrentWeatherScreen(
                viewModel = currentWeatherViewModel,
                onOpenForecast = { city ->
                    navController.navigate(Route.Forecast(city = city))
                }
            )
        }

        composable<Route.Forecast> { entry ->
            val args = entry.toRoute<Route.Forecast>()
            val city = args.city

            ForecastScreen(
                city = city,
                viewModel = forecastViewModel,
                onBackClicked = {
                    navController.popBackStack()
                }
            )
        }
    }
}